dojo/finding_group/redis.py

```diff
@@ -0,0 +1,248 @@
+                active=True,
+            ),
+        ):
+            status = "severity-Info"
+            status_text = "No SLA set, but at least one finding is active"
+            days_remaining = "No SLA"
+        else:
+            status = "age-blue"
+            status_text = "No active finding"
+            days_remaining = "Concluded"
+        title = (
+            f'<a class="has-popover" data-toggle="tooltip" data-placement="bottom" title="" href="#" data-content="{status_text}">'
+            f'<span class="label severity {status}">{days_remaining}</span></a>'
+        )
+        return mark_safe(title)
+
+
+@lru_cache(maxsize=1)
+def get_redis_client() -> redis.Redis:
+    return redis.Redis(host="redis", port=6379, decode_responses=True)
+
+
+def get_user_mode(user_id: int) -> GroupMode | None:
+    redis_client = get_redis_client()
+    value = redis_client.hget(USER_MODES_KEY, str(user_id))
+    if value and value not in [m.value for m in GroupMode]:
+        logger.warning(f"Invalid group mode '{value}' found in Redis for user {user_id}, resetting to None.")
+        redis_client.hdel(USER_MODES_KEY, str(user_id))
+        return None
+    return GroupMode(value) if value else None
+
+
+def set_user_mode(user_id: int, mode: GroupMode) -> None:
+    redis_client = get_redis_client()
+    redis_client.hset(USER_MODES_KEY, str(user_id), mode.value)
+    logger.info(f"User {user_id} dynamic finding groups mode set to {mode.value}")
+
+
+def load_or_rebuild_finding_groups(mode: GroupMode) -> dict[str, DynamicFindingGroups]:
+    redis_client = get_redis_client()
+    fg_key = DynamicFindingGroups.get_fg_key(mode)
+    id_map_key = DynamicFindingGroups.get_id_map_key(mode)
+
+    if not redis_client.exists(SYSTEM_CHANGE):
+        DynamicFindingGroups.set_last_finding_change()
+    last_finding_change_raw = redis_client.get(SYSTEM_CHANGE)
+    try:
+        last_finding_change = datetime.fromisoformat(last_finding_change_raw)
+    except ValueError:
+        logger.warning(f"Invalid datetime format in Redis for {SYSTEM_CHANGE}: {last_finding_change_raw}, resetting last finding change.")
+        DynamicFindingGroups.set_last_finding_change()
+        last_finding_change_raw = redis_client.get(SYSTEM_CHANGE)
+        last_finding_change = datetime.fromisoformat(last_finding_change_raw) if last_finding_change_raw else None
+
+    try:
+        last_update = redis_client.hget(LAST_UPDATE, mode.value)
+        last_update = datetime.fromisoformat(last_update) if last_update else None
+    except ValueError:
+        logger.warning(f"Invalid datetime format in Redis for {LAST_UPDATE}: {last_update}")
+        last_update = None
+
+    # Check if finding_groups and id_map exist in Redis
+    # Check if last update is the same as last finding change
+    # If not, rebuild them
+    if (
+        not redis_client.exists(fg_key)
+        or not redis_client.exists(id_map_key)
+        or last_update != last_finding_change
```

Comment:
It is hard to infer the semantics from these names. Can we make them more specific? Maybe try to capture when they are updated in the variables' names.